pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -31,197 +36,312 @@ const (
 	cacheLogPrefix    = "AZUREAD_PROVIDER_CACHE"
 )
 
-type azureMSGraphClient struct {
-	authResult  *customAuthResult
-	userClient  *msgraph.UsersClient
-	groupClient *msgraph.GroupsClient
+// NewMSGraphClient creates and returns a new client for accessing the Azure
+// Graph client.
+
+// It first tries to fetch the token from the refresh token, if the access token is found in the database.
+// If that fails, it tries to acquire it directly from the auth provider with the credential (application secret in Azure).
+// It also checks that the access token has the necessary permissions.
+func NewMSGraphClient(config *v32.AzureADConfig, secrets normancorev1.SecretInterface) (*AzureMSGraphClient, error) {
+	cred, err := confidential.NewCredFromSecret(config.ApplicationSecret)
+	if err != nil {
+		return nil, fmt.Errorf("could not create a cred from a secret: %w", err)
+	}
+
+	authorityURL, err := url.JoinPath(config.Endpoint, config.TenantID)
+	if err != nil {
+		return nil, fmt.Errorf("could not create token authority url: %w", err)
+	}
+
+	tokenCache := accessTokenCache{Secrets: secrets}
+	confidentialClient, err := confidential.New(authorityURL, config.ApplicationID, cred,
+		confidential.WithCache(tokenCache))
+	if err != nil {
+		return nil, fmt.Errorf("creating MS Graph client: %w", err)
+	}
+
+	graphEndpoint, err := url.JoinPath(config.GraphEndpoint, ".default")
+	if err != nil {
+		return nil, fmt.Errorf("making graph endpoint for %s: %w", config.GraphEndpoint, err)
+	}
+
+	var ar confidential.AuthResult
+	ctx := context.Background()
+	ar, err = confidentialClient.AcquireTokenSilent(ctx, []string{graphEndpoint})
+	if err != nil {
+		logrus.Infof("[%s] failed to get the access token from cache: %s", cacheLogPrefix, err)
+		logrus.Infoln("attempting to acquire the access token by credential")
+		ar, err = confidentialClient.AcquireTokenByCredential(ctx, []string{graphEndpoint})
+		if err != nil {
+			return nil, fmt.Errorf("acquiring token by credential: %w", err)
+		}
+	} else {
+		logrus.Infof("[%s] acquired token from cache", cacheLogPrefix)
+	}
+
+	authResult := getCustomAuthResult(ar)
+
+	logrus.Infof("[%s] connecting to graph endpoint: %s", providerLogPrefix, graphEndpoint)
```

Comment:
Is this going to wind-up being overly spammy in the logs?

Reply:
Yeah, info would be too spammy, I reworked all the logs to be debug.